# **Homework 2: Modern Data Structures**
Pranati Sadhu

In [5]:
# setting up system

import pandas as pd
import numpy as np
import os

In [78]:
# import data

df = pd.read_csv('/Users/pranatisadhu/Modern Data Structures/Seasons_Stats_NBA.csv')
print(df)

         Year             Player  Pos   Age   Tm     G    GS      MP   PER  \
0      1950.0    Curly Armstrong  G-F  31.0  FTW  63.0   NaN     NaN   NaN   
1      1950.0       Cliff Barker   SG  29.0  INO  49.0   NaN     NaN   NaN   
2      1950.0      Leo Barnhorst   SF  25.0  CHS  67.0   NaN     NaN   NaN   
3      1950.0         Ed Bartels    F  24.0  TOT  15.0   NaN     NaN   NaN   
4      1950.0         Ed Bartels    F  24.0  DNN  13.0   NaN     NaN   NaN   
...       ...                ...  ...   ...  ...   ...   ...     ...   ...   
24686  2017.0        Cody Zeller   PF  24.0  CHO  62.0  58.0  1725.0  16.7   
24687  2017.0       Tyler Zeller    C  27.0  BOS  51.0   5.0   525.0  13.0   
24688  2017.0  Stephen Zimmerman    C  20.0  ORL  19.0   0.0   108.0   7.3   
24689  2017.0        Paul Zipser   SF  22.0  CHI  44.0  18.0   843.0   6.9   
24690  2017.0        Ivica Zubac    C  19.0  LAL  38.0  11.0   609.0  17.0   

         TS%  ...    FT%    ORB    DRB    TRB    AST   STL   BL

# **Which NBA player scored the most points in a single season? What seasons/year was this?**

In [79]:
best_performance = df[['Year','Player','PTS']].sort_values(by = 'PTS', ascending = False)
print(best_performance)

         Year             Player     PTS
1706   1962.0  Wilt Chamberlain*  4029.0
1827   1963.0  Wilt Chamberlain*  3586.0
8469   1987.0    Michael Jordan*  3041.0
1593   1961.0  Wilt Chamberlain*  3033.0
1962   1964.0  Wilt Chamberlain*  2948.0
...       ...                ...     ...
21678     NaN                NaN     NaN
22252     NaN                NaN     NaN
22864     NaN                NaN     NaN
23516     NaN                NaN     NaN
24095     NaN                NaN     NaN

[24691 rows x 3 columns]


Wilt Chamberlain score the most points during his season in 1962. He scored 4019 points. 

# **Which NBA player scored the most points in 1991?**

In [80]:
best_performance_1991 = df[['Year','Player','PTS']].where(df['Year'] == 1991).sort_values(by = 'PTS', ascending = False)
print(best_performance_1991)

         Year              Player     PTS
10212  1991.0     Michael Jordan*  2580.0
10251  1991.0        Karl Malone*  2382.0
10132  1991.0      Patrick Ewing*  2154.0
10280  1991.0       Chris Mullin*  2107.0
10414  1991.0  Dominique Wilkins*  2101.0
...       ...                 ...     ...
24686     NaN                 NaN     NaN
24687     NaN                 NaN     NaN
24688     NaN                 NaN     NaN
24689     NaN                 NaN     NaN
24690     NaN                 NaN     NaN

[24691 rows x 3 columns]


In 1991, Micheal Jordan score the most points at 2580.

# **Which player had the best free throw percentage ("FT%") from the year 2000 to the most recent year in the data? If there are multiple players who have tied for the best percentage, then sort on the column with player names.**

In [81]:
free_throw = df[df['Year'] >= 2000][['Year','Player','FT%']].sort_values(by = 'FT%', ascending = False)
print(free_throw[free_throw['FT%']== free_throw['FT%'].max()].sort_values(by = 'Player'))

         Year           Player  FT%
17990  2006.0      Aaron Miles  1.0
16181  2003.0  Adam Harrington  1.0
22788  2014.0    Adonis Thomas  1.0
22787  2014.0    Adonis Thomas  1.0
18961  2008.0   Adrian Griffin  1.0
...       ...              ...  ...
17486  2005.0    Wesley Person  1.0
18271  2007.0     Will Blalock  1.0
17572  2005.0      Yuta Tabuse  1.0
17167  2005.0  Zarko Cabarkapa  1.0
16176  2003.0  Zendon Hamilton  1.0

[313 rows x 3 columns]


There are 313 players who tied at 100% for the highest FT% from 2000.

# **Among those with at least 50 attempts ('FTA'), which player had the best free throw percentage ("FT%") from the year 2000 to the most recent year in the data?**

In [83]:
free_throw_50 = df[(df['FTA']>= 50) & (df['Year'] >= 2000)][['Year','Player','FTA','FT%']].sort_values(by = 'FT%', ascending = False)
print(free_throw_50)

         Year           Player    FTA    FT%
15565  2002.0  Terrell Brandon   84.0  0.988
19434  2009.0    Jose Calderon  154.0  0.981
22484  2014.0    Danny Granger   53.0  0.962
22671  2014.0        Gary Neal   51.0  0.961
19350  2009.0        Ray Allen  249.0  0.952
...       ...              ...    ...    ...
14837  2000.0   Olden Polynice   90.0  0.311
17715  2006.0  Andris Biedrins   72.0  0.306
23225  2015.0      Ian Mahinmi  102.0  0.304
23026  2015.0      Joey Dorsey   83.0  0.289
15368  2001.0   Olden Polynice   65.0  0.262

[5821 rows x 4 columns]


Terrell Brandon had the highest FT% at 98.8%

# **Rename the variable "Pos" to "Position".**

In [84]:
df = df.rename(columns = {'Pos': 'Position'})
print(df)

         Year             Player Position   Age   Tm     G    GS      MP  \
0      1950.0    Curly Armstrong      G-F  31.0  FTW  63.0   NaN     NaN   
1      1950.0       Cliff Barker       SG  29.0  INO  49.0   NaN     NaN   
2      1950.0      Leo Barnhorst       SF  25.0  CHS  67.0   NaN     NaN   
3      1950.0         Ed Bartels        F  24.0  TOT  15.0   NaN     NaN   
4      1950.0         Ed Bartels        F  24.0  DNN  13.0   NaN     NaN   
...       ...                ...      ...   ...  ...   ...   ...     ...   
24686  2017.0        Cody Zeller       PF  24.0  CHO  62.0  58.0  1725.0   
24687  2017.0       Tyler Zeller        C  27.0  BOS  51.0   5.0   525.0   
24688  2017.0  Stephen Zimmerman        C  20.0  ORL  19.0   0.0   108.0   
24689  2017.0        Paul Zipser       SF  22.0  CHI  44.0  18.0   843.0   
24690  2017.0        Ivica Zubac        C  19.0  LAL  38.0  11.0   609.0   

        PER    TS%  ...    FT%    ORB    DRB    TRB    AST   STL   BLK   TOV  \
0      

# **Use this variable to create two variables that are called "first_position" and "second_position". Hint: If a player plays two positions, then the "-" is used as a delimiter between their first and second positions. In these cases, you will have to split the position variable into two.**

In [85]:
df['first_position'] = df['Position'].str.split('-').str[0]
df['second_position'] = df['Position'].str.split('-').str[1]
print(df[['Position', 'first_position', 'second_position']])

      Position first_position second_position
0          G-F              G               F
1           SG             SG             NaN
2           SF             SF             NaN
3            F              F             NaN
4            F              F             NaN
...        ...            ...             ...
24686       PF             PF             NaN
24687        C              C             NaN
24688        C              C             NaN
24689       SF             SF             NaN
24690        C              C             NaN

[24691 rows x 3 columns]


# **Combine these two variables back into a single variable called "position_united". Make sure to separate the two positions by "-", as the original data field did. If there is no second position, then there will only be a single position. Again, this is similar to how the original variable was coded. Present a frequency table of "position_united".**

In [86]:
df['position_united'] = df.apply(lambda row : row['first_position'] + '-' + row['second_position'] if pd.notna(row['second_position']) else row['first_position'], axis = 1)
print(df['position_united'].value_counts())

position_united
PF       4966
SG       4811
C        4759
SF       4699
PG       4648
G         139
F         123
F-C        81
G-F        79
F-G        60
C-F        35
C-PF       33
SF-SG      30
PG-SG      28
SG-PG      27
PF-C       27
SG-SF      26
PF-SF      25
SF-PF      21
SG-PF       3
C-SF        2
SF-PG       1
PG-SF       1
Name: count, dtype: int64


# **Create two new datasets.**
- a new dataset from the original dataset that includes all data except the age variable (be sure to give this dataset a new name). 
- a new dataset from the original dataset that includes the year, the player name, and age.
- add a new column to both datasets called mergeid that includes a sequence of numbers beginning with a 1 in the first row of the data and ending with the total number of rows in the last row of the data

In [87]:
df_ageless = df.drop(columns = 'Age').copy()
print(df_ageless)

         Year             Player Position   Tm     G    GS      MP   PER  \
0      1950.0    Curly Armstrong      G-F  FTW  63.0   NaN     NaN   NaN   
1      1950.0       Cliff Barker       SG  INO  49.0   NaN     NaN   NaN   
2      1950.0      Leo Barnhorst       SF  CHS  67.0   NaN     NaN   NaN   
3      1950.0         Ed Bartels        F  TOT  15.0   NaN     NaN   NaN   
4      1950.0         Ed Bartels        F  DNN  13.0   NaN     NaN   NaN   
...       ...                ...      ...  ...   ...   ...     ...   ...   
24686  2017.0        Cody Zeller       PF  CHO  62.0  58.0  1725.0  16.7   
24687  2017.0       Tyler Zeller        C  BOS  51.0   5.0   525.0  13.0   
24688  2017.0  Stephen Zimmerman        C  ORL  19.0   0.0   108.0   7.3   
24689  2017.0        Paul Zipser       SF  CHI  44.0  18.0   843.0   6.9   
24690  2017.0        Ivica Zubac        C  LAL  38.0  11.0   609.0  17.0   

         TS%   3PAr  ...    TRB    AST   STL   BLK   TOV     PF    PTS  \
0      0.368 

In [88]:
df_condensed = df[['Year','Player','Age']].copy()
print(df_condensed)

         Year             Player   Age
0      1950.0    Curly Armstrong  31.0
1      1950.0       Cliff Barker  29.0
2      1950.0      Leo Barnhorst  25.0
3      1950.0         Ed Bartels  24.0
4      1950.0         Ed Bartels  24.0
...       ...                ...   ...
24686  2017.0        Cody Zeller  24.0
24687  2017.0       Tyler Zeller  27.0
24688  2017.0  Stephen Zimmerman  20.0
24689  2017.0        Paul Zipser  22.0
24690  2017.0        Ivica Zubac  19.0

[24691 rows x 3 columns]


In [89]:
df_ageless['mergeid'] = df_ageless.index + 1
df_condensed['mergeid'] = df_condensed.index + 1
print("Data Frame without the Age Column (mergeid added):\n")
print(df_ageless)
print("Condensed Data Frame (mergeid added):\n")
print(df_condensed)

Data Frame without the Age Column (mergeid added):

         Year             Player Position   Tm     G    GS      MP   PER  \
0      1950.0    Curly Armstrong      G-F  FTW  63.0   NaN     NaN   NaN   
1      1950.0       Cliff Barker       SG  INO  49.0   NaN     NaN   NaN   
2      1950.0      Leo Barnhorst       SF  CHS  67.0   NaN     NaN   NaN   
3      1950.0         Ed Bartels        F  TOT  15.0   NaN     NaN   NaN   
4      1950.0         Ed Bartels        F  DNN  13.0   NaN     NaN   NaN   
...       ...                ...      ...  ...   ...   ...     ...   ...   
24686  2017.0        Cody Zeller       PF  CHO  62.0  58.0  1725.0  16.7   
24687  2017.0       Tyler Zeller        C  BOS  51.0   5.0   525.0  13.0   
24688  2017.0  Stephen Zimmerman        C  ORL  19.0   0.0   108.0   7.3   
24689  2017.0        Paul Zipser       SF  CHI  44.0  18.0   843.0   6.9   
24690  2017.0        Ivica Zubac        C  LAL  38.0  11.0   609.0  17.0   

         TS%   3PAr  ...    AST   S

# **Join the two datasets from question (8) together to recreate the original dataset plus the new merge id.**

In [91]:
merged_df = df_ageless.merge(df_condensed, on = 'mergeid', how = 'inner')
cols_to_drop = [col for col in merged_df.columns if col.endswith('_y')]
merged_df = merged_df.drop(columns=cols_to_drop)
merged_df.columns = merged_df.columns.str.replace('_x', '')
print(merged_df)

         Year             Player Position   Tm     G    GS      MP   PER  \
0      1950.0    Curly Armstrong      G-F  FTW  63.0   NaN     NaN   NaN   
1      1950.0       Cliff Barker       SG  INO  49.0   NaN     NaN   NaN   
2      1950.0      Leo Barnhorst       SF  CHS  67.0   NaN     NaN   NaN   
3      1950.0         Ed Bartels        F  TOT  15.0   NaN     NaN   NaN   
4      1950.0         Ed Bartels        F  DNN  13.0   NaN     NaN   NaN   
...       ...                ...      ...  ...   ...   ...     ...   ...   
24686  2017.0        Cody Zeller       PF  CHO  62.0  58.0  1725.0  16.7   
24687  2017.0       Tyler Zeller        C  BOS  51.0   5.0   525.0  13.0   
24688  2017.0  Stephen Zimmerman        C  ORL  19.0   0.0   108.0   7.3   
24689  2017.0        Paul Zipser       SF  CHI  44.0  18.0   843.0   6.9   
24690  2017.0        Ivica Zubac        C  LAL  38.0  11.0   609.0  17.0   

         TS%   3PAr  ...   STL   BLK   TOV     PF    PTS  first_position  \
0      0.36

# **Subset the original dataset to the years 1995-1997. Group the data by team name and year and then summarize the average number of points per team. Arrange from most to least points.**

In [92]:
df[(df['Year'] >= 1995) & (df['Year'] <= 1997)][['Tm', 'Year', 'PTS']].groupby(['Tm', 'Year']).mean('PTS').sort_values(by = 'PTS', ascending = False)

PTS
Tm  Year              
SEA 1996.0  659.384615
    1995.0  646.785714
MIN 1997.0  606.307692
ORL 1995.0  606.066667
PHO 1995.0  604.866667
...                ...
DEN 1997.0  348.695652
NJN 1997.0  346.695652
PHI 1996.0  322.750000
TOT 1997.0  322.380952
DAL 1997.0  275.222222

[88 rows x 1 columns]

# **Similar to Q10, start by subsetting the original dataset to the years 1995-1997. Then create a wide dataset that keeps each year in its own row, but spreads team names to multiple columns with each column delineating points per team in each of the 1995-1997 seasons (Hint: Your final answer should have a shape of 3 x 30).**

In [93]:
df_wider = df[(df['Year'] >= 1995) & (df['Year'] <= 1997)][['Tm','Year','PTS']].pivot_table(index = 'Year', columns = 'Tm', values='PTS', aggfunc = 'mean')
df_wider

Tm,ATL,BOS,CHH,CHI,CLE,DAL,DEN,DET,GSW,HOU,...,PHO,POR,SAC,SAS,SEA,TOR,TOT,UTA,VAN,WSB
Year,,,,,,,,,,,,,,,,,,,,,
1995.0,440.055556,495.764706,515.562500,520.312500,494.466667,604.428571,488.764706,503.312500,481.500000,499.470588,...,604.866667,528.187500,537.066667,546.375000,646.785714,NaN,488.681818,581.733333,NaN,515.125000
1996.0,447.722222,471.944444,401.476190,575.000000,498.200000,560.600000,471.352941,558.714286,490.235294,466.888889,...,503.058824,543.000000,510.187500,565.133333,659.384615,420.736842,419.767857,600.285714,409.0000,400.380952
1997.0,457.294118,458.222222,450.444444,563.866667,551.769231,275.222222,348.695652,482.687500,510.687500,458.222222,...,366.565217,427.052632,494.250000,370.900000,517.125000,372.809524,322.380952,563.600000,457.0625,543.133333


# **Now return the data from #11 back into a long (tidy) format by moving Team (TM), Points and Year into their own columns**

In [94]:
df_longer = df_wider.reset_index().melt(id_vars = ['Year'], var_name = 'Tm',value_name = 'PTS')
df_longer

,Year,Tm,PTS
0,1995.0,ATL,440.055556
1,1996.0,ATL,447.722222
2,1997.0,ATL,457.294118
3,1995.0,BOS,495.764706
4,1996.0,BOS,471.944444
...,...,...,...
85,1996.0,VAN,409.000000
86,1997.0,VAN,457.062500
87,1995.0,WSB,515.125000
88,1996.0,WSB,400.380952
